### PR TITLE
Heterogeneous API - backend independent C-API for ArrayFire

### PR DESCRIPTION
**Pull request by 9prady9**

### CODE REVIEW

**Comment by pavanky** [comment]
@9prady9 For now manually include `src/api/c/seq.cpp` while compiling hapi. This solves the issue.

File: CMakeLists.txt [hunk 1]
```diff
@@ -168,6 +170,11 @@ IF(${BUILD_OPENCL})
     ADD_SUBDIRECTORY(src/backend/opencl)
 ENDIF()
 
+IF(${BUILD_HETEROGENOUS_API})
+    ADD_SUBDIRECTORY(src/api/hapi)
+    ADD_SUBDIRECTORY(hapi_examples)
```
Comment (pavanky) [inline CMakeLists.txt]: I don't think the examples need to exist in different directories. We can add dummy variations of all hapi specific functions in each backend.

File: hapi_examples/test.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    af_set_backend(AF_BACKEND_OPENCL);
+    testBackend();
+
+    #ifdef WIN32 // pause in Windows
+    if (!(argc == 2 && argv[1][0] == '-')) {
+        printf("hit [enter]...");
+        fflush(stdout);
+        getchar();
+    }
+    #endif
```
Comment (pavanky) [inline hapi_examples/test.cpp]: Why not use the C++ API ? We should have `setBackend()` in C++ as well.

File: include/af/defines.h [hunk 3]
```diff
@@ -316,6 +321,12 @@ typedef enum {
     AF_FIF_RAW          = 34    ///< FreeImage Enum for RAW Camera Image File
 } af_image_format;
 
+typedef enum {
+    AF_BACKEND_CPU,
+    AF_BACKEND_CUDA,
+    AF_BACKEND_OPENCL
```
Comment (pavanky) [inline include/af/defines.h]: Add values.

File: include/af/hapi.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ ********************************************************/
+
+#pragma once
+#include <af/defines.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+AFAPI af_err af_set_backend(const af_backend bknd);
```
Comment (pavanky) [inline include/af/hapi.h]: Document this

File: src/api/hapi/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    "${CMAKE_CURRENT_SOURCE_DIR}/*.cpp")
+
+ADD_LIBRARY(af SHARED
+            ${hapi_headers}
+            ${hapi_sources})
+
+IF(${BUILD_CPU})
+    ADD_DEPENDENCIES(af afcpu)
+ENDIF()
+
```
Comment (pavanky) [inline src/api/hapi/CMakeLists.txt]: Technically hapi does not need to wait for any backend to be built. There is no reason for this.

File: src/api/hapi/symbol_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    if (cudaHandle) {
+        isCUDALoaded = true;
+        activeHandle = cudaHandle;
+    }
+    oclHandle = openDynLibrary(LIB_AF_OCL_NAME);
+    if (oclHandle) {
+        isOCLLoaded = true;
+        activeHandle = oclHandle;
+    }
+}
```
Comment (pavanky) [inline src/api/hapi/symbol_manager.cpp]: We need to figure out the preferred order or atleast provide the user with a choice. Right now it defaults to OpenCL > CUDA > CPU.

File: src/api/hapi/symbol_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        isOCLLoaded = false;
+    }
+}
+
+af_err AFSymbolManager::setBackend(af::Backend bknd)
+{
+    af_err retCode = AF_SUCCESS;
+    activeBknd = bknd;
+    switch (activeBknd) {
+        case af::Backend::AF_BACKEND_CPU:
```
Comment (pavanky) [inline src/api/hapi/symbol_manager.cpp]: If you make `af_backend` an enum with pre defined numbers you can have a lookup table like `isBackendLoaded[kind]`

File: src/api/hapi/symbol_manager.hpp [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            funcHandle = (af_func)GetProcAddress(activeHandle, symbolName);
+#else
+            funcHandle = (af_func)dlsym(activeHandle, symbolName);
+#endif
+            if (!funcHandle) {
+                return AF_ERR_SYM_LOAD;
+            }
+            return funcHandle(args...);
+        }
+
```
Comment (pavanky) [inline src/api/hapi/symbol_manager.hpp]: This function has to be the best thing about this entire process :D

File: src/api/hapi/symbol_manager.hpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    private:
+        bool isCPULoaded;
+        bool isCUDALoaded;
+        bool isOCLLoaded;
+
+        LibHandle cpuHandle;
+        LibHandle cudaHandle;
+        LibHandle oclHandle;
+
```
Comment (pavanky) [inline src/api/hapi/symbol_manager.hpp]: Read what I said earlier about lookup tables

**Comment by pavanky** [comment]
@9prady9 good work! This went faster than I thought. Are the tests passing on the dev machines?

File: CMakeLists.txt [hunk 10]
```diff
@@ -168,6 +170,11 @@ IF(${BUILD_OPENCL})
     ADD_SUBDIRECTORY(src/backend/opencl)
 ENDIF()
 
+IF(${BUILD_HETEROGENOUS_API})
```
Comment (pavanky) [inline CMakeLists.txt]: Change this be "BUILD_AF"

File: include/af/defines.h (hunk 3 above)
Comment (umar456) [inline include/af/defines.h]: Also add AF_BACKEND_DEFAULT which will pick CUDA->OpenCL -> CPU

File: hapi_examples/test.cpp (hunk 2 above)
Comment (9prady9) [inline hapi_examples/test.cpp]: Sure, it's just that i just worked on only C-API right now.

File: include/af/defines.h (hunk 3 above)
Comment (9prady9) [inline include/af/defines.h]: That's the behavior anyway right except that there is no specific enum for that.

File: src/api/hapi/CMakeLists.txt (hunk 5 above)
Comment (9prady9) [inline src/api/hapi/CMakeLists.txt]: Sure, hapi is just an c++ library that will use af*.so's. We can remove the dependency.

File: src/api/hapi/symbol_manager.cpp (hunk 6 above)
Comment (9prady9) [inline src/api/hapi/symbol_manager.cpp]: I didn't quite understand what is the choice here ? isn't that the reason af_set_backend function is there. In the even of such function is not called upon, activeHandle points to the backend in the order you just mentioned.

File: src/api/hapi/symbol_manager.hpp (hunk 8 above)
Comment (9prady9) [inline src/api/hapi/symbol_manager.hpp]: true.

File: src/api/hapi/symbol_manager.cpp (hunk 7 above)
Comment (9prady9) [inline src/api/hapi/symbol_manager.cpp]: agree.

**Comment by 9prady9** [comment]
@pavanky All builds passed. I think we can merge this if there are no more additions or missing functionality.

**Comment by pavanky** [comment]
Wait until #975 is merged into devel, merge from devel and add those functions as well.

**Comment by pavanky** [comment]
@9prady9 
Can you add the list of new functions added in the description of this issue: https://github.com/arrayfire/arrayfire/issues/984

File: include/af/hapi.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
+/*******************************************************
```
Comment (pavanky) [inline include/af/hapi.h]: I think this file should now be called `af/backend.h`

**Comment by pavanky** [comment]
Tests are going to fail if all tests are being run instead of specifying each backend separately using `ctest -R backend`. This is because `libaf.so` does not currently know the location of each individual file at the time of the test.

**Comment by pavanky** [comment]
There is also the following error that occurs at the exit of each program. I am not sure if this is limited to `belle`.

```
 $ ./build/test/info_unified 
Running main() from gtest_main.cc
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Info/0, where TypeParam = float
[ RUN      ] Info/0.All
ERROR: GLFW wasn't able to initalize
ArrayFire v3.1.0 (OpenCL, 64-bit Linux, build 4cb64cb)
[0] NVIDIA  : Tesla K20c 
-1- NVIDIA  : Quadro 6000 
-2- NVIDIA  : Tesla K20c 
ArrayFire v3.1.0 (OpenCL, 64-bit Linux, build 4cb64cb)
-0- NVIDIA  : Tesla K20c 
[1] NVIDIA  : Quadro 6000 
-2- NVIDIA  : Tesla K20c 
ArrayFire v3.1.0 (OpenCL, 64-bit Linux, build 4cb64cb)
-0- NVIDIA  : Tesla K20c 
-1- NVIDIA  : Quadro 6000 
[2] NVIDIA  : Tesla K20c 
[       OK ] Info/0.All (699 ms)
[----------] 1 test from Info/0 (699 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (699 ms total)
[  PASSED  ] 1 test.

[ FATAL ] /home/pavan/Work/arrayfire/test/gtest/include/gtest/internal/gtest-port.h:2019:: pthread_key_delete(key_)failed with error 22
Aborted (core dumped)
```

**Comment by pavanky** [comment]
@shehzan10 @umar456 @9prady9 

One way to deal with tests would be to use a priority list of paths for loading the libraries instead of looking for absolute paths. This includes something like the following.

1) `libaf$BACKEND`
2) `$AF_PATH/lib/libaf$BACKEND`
3) `$AF_BUILD_PATH/src/build/lib$BACKEND`

SO when running tests we could run using something like

`AF_BUILD_PATH=/full/path/to/af/build ctest -R unified`

**Comment by 9prady9** [comment]
@pavanky @umar456  @shehzan10 If anyone of you also tried the latest changes, most of the functions passed tests on `unified` category as well. Exceptions are given below
* sort_by_index_unified
* sort_index_unified
* blas_unified
* imageio_unified

Both sort ones failed with value errors where as the other two were seg faults. Are you seeing anything similar on your machine too ?

**Comment by 9prady9** [comment]
@pavanky I have just pushed the changes related to search path order you have suggested earlier. Please take a look at them.

**Comment by pavanky** [comment]
@9prady9 I am not sure how the `*_unified` tests were even running. The dynamic loader was perhaps picking up some other files?

**Comment by 9prady9** [comment]
Looks like there are some issues on windows.
Didn't quite understand your last comment, we are not overriding any system path library ? PATH or LD_LIBRARY_PATH is the first place the library is searched for.

**Comment by 9prady9** [comment]
@pavanky What should we do about the failing tests in unified category ?

**Comment by 9prady9** [comment]
Figured out the fix for imageio test. However, i am still perplexed why sort tests are failing.

**Comment by 9prady9** [comment]
build arrayfire ci

**Pull request merged**